Updating the api chunking proposal to add remainingItemCount

caesarxuchao opened the pull request:
/sig api-machinery

Retrospectively updating the proposal for api-chunking for the remainingItemCount API change introduced in https://github.com/kubernetes/kubernetes/pull/75993.

/assign @smarterclayton 
cc @lavalamp

smarterclayton commented on contributors/design-proposals/api-machinery/api-chunking.md:
```diff
@@ -78,6 +80,7 @@ The server **may** limit the amount of time a continue token is valid for. Clien
 
 The server **must** support `continue` tokens that are valid across multiple API servers. The server **must** support a mechanism for rolling restart such that continue tokens are valid after one or all API servers have been restarted.
 
+The `remainingItemCount` returned by the server is the number of subsequent items in the list which are not included in this list response. If the list request contained label or field selectors, then the number of remaining items is unknown and this field will be unset and omitted during serialization. If the list is complete (either because the list request is not a chunking one or because this ...
```
Is it required to be exact?  What’s the use case?

smarterclayton replied:
What happens when I’m halfway through a list?

smarterclayton commented on contributors/design-proposals/api-machinery/api-chunking.md:
```diff
@@ -91,6 +94,8 @@ Implementations that cannot offer consistent ranging (returning a set of results
 
 For etcd3 the continue token would contain a resource version (the snapshot that we are reading that is consistent across the entire LIST) and the start key for the next set of results. Upon receiving a valid continue token the apiserver would instruct etcd3 to retrieve the set of results at a given resource version, beginning at the provided start key, limited by the maximum number of requests p ...
 
+etcd3 returns the total number of keys within the range as `response.Count` if the read is a range read. The storage checks if the list request contained label or field selector. If not, the storage calculates the `remainingItemCount` as `response.Count-limit`.
```
If so, what happens?

caesarxuchao commented:
@smarterclayton I added use cases and why we needed the field. PTAL.

The remainingItemCount is exact. The doc never suggests that the count is an approximation, so I think readers will get it.

smarterclayton commented:
> The remainingItemCount is exact. The doc never suggests that the count is an approximation, so I think readers will get it.

That's actually my concern, that we enshrine an exact count as part of the API.  I would generally want to be as vague as possible to leave open the door for either relaxed semantics (performance) or alternate backends (proxies that can't necessarily know).

smarterclayton commented:
Also add what feature flag gate this will be under and what release it will be promoted (1.16 is ok for promotion to GA if we get good data in 1.15).

smarterclayton commented on contributors/design-proposals/api-machinery/api-chunking.md:
```diff
@@ -78,6 +80,11 @@ The server **may** limit the amount of time a continue token is valid for. Clien
+
+For example, assuming there are 1450 pods in the cluster when the client sends a chunking list request for pods, with `limit` set to 500. The first chunked list response contains 500 pods, a `continue` token, and with `metadata.remainingItemCount` set to 950. If the client use the `continue` token to continue listing, the server returns the second chunked list response containing the next 500 pod ...
+
+The `remainingItemCount` offers a simple and efficient way to get the count of objects of a resource type. For example, to get the total count of all pods in the cluster, simply sends `GET /api/v1/pods?limit=1` to the apiserver. Similarly, one can get the total count of all pods in a namespace. Note that the `remainingItemCount` is set to 0 when the list request contains any label or field select ...
```
Since old servers do not support this, and I don't think we can require this value in all possible future implementations, you need to specifically describe to a client how they can tell when the server doesn't support this field, and instruct them how to correctly detect it (as other sections do).

Effectively the algorithm is:

1. Make a limit request
2. Observe whether continue is set (if not, you have only one or zero items)
3. Observe whether remainingItemCount is set (if not, remainingItemCount isn't supported)

smarterclayton commented:
Hrm, I didn't see a description of the use cases in the form I would expect.  I.e. "we want to let people estimate the size of the collection", "we want to make this part of a UI" (I hope it's the former rather than the latter, because we discouraged use of chunking for human UI viewing).

lavalamp commented:
Yeah, it's the former, specifically we want to estimate when the storage
migrator will finish a migration.

*From: *Clayton Coleman <notifications@github.com>
*Date: *Tue, May 14, 2019 at 7:56 AM
*To: *kubernetes/community
*Cc: *Daniel Smith, Mention

Hrm, I didn't see a description of the use cases in the form I would
> expect. I.e. "we want to let people estimate the size of the collection",
> "we want to make this part of a UI" (I hope it's the former rather than the
> latter, because we discouraged use of chunking for human UI viewing).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/3710?email_source=notifications&email_token=AAE6BFTIC6WM77AAYMEV7DTPVLHIDA5CNFSM4HMU6A3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVLX5LA#issuecomment-492273324>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAE6BFWPBD4GPFOEPSJIHALPVLHIDANCNFSM4HMU6A3A>
> .
>

caesarxuchao commented:
RE. concerns over the "exact count", I added "The intended use of the `remainingItemCount` is *estimating* the size of a collection. Clients should not rely on the `remainingItemCount` to be set or to be exact."

RE. detecting if `remainingItemCount` is supported, I added instructions.

RE. feature flag, I'll wait for you and Daniel to converge. See Daniel's comment at  https://github.com/kubernetes/kubernetes/pull/75993#issuecomment-492308555. Starting the field as beta sounds more useful to me.

RE. use cases, perhaps I didn't understand your request, let me know if the current wording is still too vague.

caesarxuchao commented:
Added feature flag. @smarterclayton PTAL. Thank you.
